streaming/out.go

```diff
@@ -75,6 +76,72 @@ func NewCrypt4GHWriter(writer io.Writer, writerPrivateKey [chacha20poly1305.KeyS
 	return &crypt4GHWriter, nil
 }
 
+func NewReCrypt4GHWriter(reader io.Reader, writer io.Writer, writerPrivateKey [chacha20poly1305.KeySize]byte, readerPublicKeyList [][chacha20poly1305.KeySize]byte, trim bool) (*Crypt4GHWriter, error) {
+	crypt4GHWriter := Crypt4GHWriter{}
+	binaryHeader, err := headers.ReadHeader(reader)
+	if err != nil {
+		return nil, err
+	}
+	crypt4GHInternalReader := crypt4GHInternalReader{header: make([]byte, len(binaryHeader))}
+	copy(crypt4GHInternalReader.header, binaryHeader)
+	// buffer := bytes.NewBuffer(binaryHeader)
+	// header, err := headers.NewHeader(buffer, writerPrivateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	var sharedKey [chacha20poly1305.KeySize]byte
+	_, err = rand.Read(sharedKey[:])
+	if err != nil {
+		return nil, err
+	}
+
+	headerPackets := make([]headers.HeaderPacket, 0)
+	crypt4GHWriter.dataEncryptionParametersHeaderPacket = headers.DataEncryptionParametersHeaderPacket{
+		EncryptedSegmentSize: chacha20poly1305.NonceSize + headers.UnencryptedDataSegmentSize + box.Overhead,
+		PacketType:           headers.PacketType{PacketType: headers.DataEncryptionParameters},
+		DataEncryptionMethod: headers.ChaCha20IETFPoly1305,
+		DataKey:              sharedKey,
+	}
+
+	for _, readerPublicKey := range readerPublicKeyList {
+		headerPackets = append(headerPackets, headers.HeaderPacket{
+			WriterPrivateKey:       writerPrivateKey,
+			ReaderPublicKey:        readerPublicKey,
+			HeaderEncryptionMethod: headers.X25519ChaCha20IETFPoly1305,
+			EncryptedHeaderPacket:  crypt4GHWriter.dataEncryptionParametersHeaderPacket,
+		})
+	}
+
+	headerPackets = append(headerPackets, headers.HeaderPacket{
+		WriterPrivateKey:       writerPrivateKey,
+		ReaderPublicKey:        keys.DerivePublicKey(writerPrivateKey),
+		HeaderEncryptionMethod: headers.X25519ChaCha20IETFPoly1305,
+		EncryptedHeaderPacket:  crypt4GHWriter.dataEncryptionParametersHeaderPacket,
+	})
+
+	var magicNumber [8]byte
+	copy(magicNumber[:], headers.MagicNumber)
+	crypt4GHWriter.header = headers.Header{
+		MagicNumber:       magicNumber,
+		Version:           headers.Version,
+		HeaderPacketCount: uint32(len(headerPackets)),
+		HeaderPackets:     headerPackets,
+	}
+
+	newBinaryHeader, err := crypt4GHWriter.header.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	_, err = writer.Write(newBinaryHeader)
+	if err != nil {
+		return nil, err
+	}
+	crypt4GHWriter.writer = writer
+	crypt4GHWriter.buffer.Grow(headers.UnencryptedDataSegmentSize)
+	return &crypt4GHWriter, nil
+}
+
```

Comment:
This function should have a test case

Reply:
do I add it as part of this PR or another?

Reply:
The best would be to put this and the corresponding test in a separate PR.
That way this PR don't introduce any real code changes.

Reply:
ah my bad i think i added some changes i made for myself, will fix it